### PR TITLE
Add `#gallery` to `Spree::Variant` and `Spree::Product`

### PR DESCRIPTION
**Pull request by swcraig**
This is a lightweight version of #625.

Related to #2324, I think this is a good starting point to begin opening up an interface for different image upload implementations.

### CODE REVIEW

File: frontend/app/views/spree/products/_thumbnails.html.erb [hunk 1]
```diff
@@ -1,19 +1,10 @@
 <%# no need for thumbnails unless there is more than one image %>
```
Comment (jhawthorn) [inline frontend/app/views/spree/products/_thumbnails.html.erb]: We can't change the behaviour of this file. We can't have it so users upgrade and the logic for displaying images on the frontend changes (I think it's likely in a ton of stores that this wouldn't be caught before production).

File: core/app/models/spree/variant.rb [hunk 2]
```diff
@@ -350,6 +350,15 @@ def variant_properties
       end.flatten.compact
     end
 
+    # An enumerable of media that goes along with a variant
+    # Stores may not necessarily want to attach images to variants
+    # This provides a more robust interface
+    #
+    # @return [Enumerable] of media for a variant
+    def gallery
+      Variant.reflect_on_association(:images) ? association(:images).reader : []
```
Comment (jhawthorn) [inline core/app/models/spree/variant.rb]: We define this association, so we shouldn't test for it. If this is to prepare for a future paperclip removal PR, it should be done later.

File: core/app/models/spree/product.rb [hunk 3]
```diff
@@ -280,6 +280,19 @@ def find_variant_property_rule(option_value_ids)
       end
     end
 
+    # An enumerable of media that goes along with a product
+    # Stores may not necessarily want to attach images to products (variants)
+    # This provides a more robust interface
+    #
+    # @return [Enumerable] of media for a product (and its variants)
+    def gallery
+      if Product.method_defined?(:variant_images)
```
Comment (jhawthorn) [inline core/app/models/spree/product.rb]: We define this association, so we shouldn't test for it. If this is to prepare for a future paperclip removal PR, it should be done later.

**Comment by swcraig** [comment]
Closed in favour of #2337

**Pull request closed without merging**